File: cpp/src/compact.cpp
#include "rohrpost/compact.hpp"

#include "rohrpost/config.hpp"
#include "rohrpost/errors.hpp"
#include "rohrpost/fold.hpp"
#include "rohrpost/ids.hpp"
#include "rohrpost/io.hpp"
#include "rohrpost/paths.hpp"
#include "rohrpost/pyfmt.hpp"
#include "rohrpost/store.hpp"
#include "rohrpost/subprocess.hpp"
#include "rohrpost/timeutil.hpp"

#include <algorithm>
#include <chrono>
#include <format>
#include <map>
#include <set>

namespace rp::compact {
namespace {

namespace fs = std::filesystem;

std::int64_t parse_ts(const std::string& ts) {
    const auto parsed = timeutil::parse_ts(ts);
    if (!parsed) throw StoreError(std::format("Invalid isoformat string: {}", py::repr(ts)));
    return *parsed;
}

/// `(is_dirty, current_branch)`; `(false, "")` without git.
std::pair<bool, std::string> git_state(const fs::path& repo_root) {
    const std::string root = io::path_str(repo_root);
    auto status = subprocess::run({"git", "-C", root, "status", "--porcelain"}, std::chrono::seconds(10));
    if (!status) return {false, ""};
    const bool dirty = !py::strip(status->stdout_bytes).empty();
    auto branch = subprocess::run({"git", "-C", root, "rev-parse", "--abbrev-ref", "HEAD"}, std::chrono::seconds(10));
    if (!branch) return {false, ""};
    return {dirty, std::string(py::strip(branch->stdout_bytes))};
}

/// A refusal reason if compaction must not proceed, else nullopt.
std::optional<std::string> guard(const fs::path& repo_root, bool force, const std::optional<std::string>& default_branch) {
    if (force) return std::nullopt;
    const auto [dirty, branch] = git_state(repo_root);
    if (branch.empty()) return std::nullopt;  // outside git — nothing to protect
    if (dirty) return "refusing to compact: working tree is dirty (use --force to override)";
    const std::string expected = default_branch.value_or("main");
    if (branch != expected) {
        return std::format("refusing to compact: HEAD is on {}, not {} (use --force to override)", py::repr(branch), py::repr(expected));
    }
    return std::nullopt;
}

std::string ticket_of(const Event& ev) {
    try {
        return ids::normalize_id(ev.ticket);
    } catch (const StoreError&) {
        return ev.ticket;  // mirrors the reference: IdError propagates
    }
}

bool is_terminal_set(const Event& e) {
    if (e.op != "set" || !e.set || e.set->empty()) return false;
    const auto it = e.set->find("status");
    return it != e.set->end() && it->is_string() && is_terminal(it->get<std::string>());
}

std::set<std::string> archivable_ids(const std::vector<Event>& events, const TicketMap& by_id, std::int64_t cutoff_ms) {
    std::set<std::string> out;
    for (const auto& [tid, ticket] : by_id) {
        if (!is_terminal(ticket.status)) continue;
        std::optional<std::int64_t> latest;
        for (const auto& e : events) {
            if (is_terminal_set(e) && ticket_of(e) == tid) {
                const std::int64_t t = parse_ts(e.ts);
                if (!latest || t > *latest) latest = t;
            }
        }
        if (latest && *latest < cutoff_ms) out.insert(tid);
    }
    return out;
}

void sort_events(std::vector<Event>& events) {
    std::stable_sort(events.begin(), events.end(), [](const Event& a, const Event& b) {
        return a.ts != b.ts ? a.ts < b.ts : a.id < b.id;
    });
}

void rewrite_log(const fs::path& dir, std::vector<Event> keep) {
    sort_events(keep);
    std::string payload;
    for (const auto& e : keep) {
        payload += encode(e);
        payload.push_back('\n');
    }
    const fs::path log = paths::log_path(dir);
    fs::path tmp = log;
    tmp.replace_extension(".jsonl.tmp");
    io::write_file_atomic(log, tmp, payload);
}

void append_archive(const fs::path& dir, const std::string& bucket, std::vector<Event> evs) {
    const fs::path adir = paths::archive_dir(dir);
    std::error_code ec;
    fs::create_directories(adir, ec);
    sort_events(evs);
    std::string payload;
    for (const auto& e : evs) {
        payload += encode(e);
        payload.push_back('\n');
    }
    io::append_file(adir / io::path_from_utf8(bucket), payload);
}

void fail(const std::string& message, bool json_output) {
    if (json_output) {
        Json obj = Json::object();
        obj["error"] = message;
        io::println(json::dumps(obj, json::kPretty));
    } else {
        io::eprintln(std::format("rp compact: {}", message));
    }
}

void report(const CompactResult& result, bool json_output) {
    if (json_output) {
        io::println(json::dumps(result.to_mapping(), json::kPretty));
        return;
    }
    io::println(std::format("Compacted: archived {} event(s), kept {}.", result.archived, result.remaining));
    if (!result.archive_files.empty()) {
        std::string joined;
        for (std::size_t i = 0; i < result.archive_files.size(); ++i) {
            if (i) joined += ", ";
            joined += result.archive_files[i];
        }
        io::println("  archive files: " + joined);
    }
}

}  // namespace

Json CompactResult::to_mapping() const {
    Json m = Json::object();
    m["archived"] = archived;
    m["remaining"] = remaining;
    m["archive_files"] = archive_files;
    return m;
}

std::string quarter_bucket(const std::string& ts) {
    const auto civil = timeutil::to_civil(parse_ts(ts));
    const unsigned quarter = (civil.month - 1) / 3 + 1;
    return std::format("log-{}-Q{}.jsonl", civil.year, quarter);
}

int run(const fs::path& dir, const Options& options) {
    const Config config = load_config(dir);
    if (const auto refusal = guard(dir.parent_path(), options.force, config.default_branch)) {
        fail(*refusal, options.json_output);
        return 1;
    }
    const std::int64_t now_ms = options.now_ms.value_or(timeutil::now_epoch_ms());
    const std::int64_t cutoff = now_ms - static_cast<std::int64_t>(options.archive_after_days) * 86'400'000LL;

    std::vector<Event> keep;
    std::map<std::string, std::vector<Event>> buckets;  // sorted by bucket name
    {
        // Read and partition while holding the same lock as the rewrite.
        store::FileLock lock(dir);
        const std::vector<Event> events_all = store::read_events(dir);
        const TicketMap by_id = fold(events_all);
        const auto archivable = archivable_ids(events_all, by_id, cutoff);
        for (const auto& ev : events_all) {
            if (archivable.contains(ticket_of(ev))) buckets[quarter_bucket(ev.ts)].push_back(ev);
            else keep.push_back(ev);
        }
        rewrite_log(dir, keep);
        for (auto& [bucket, evs] : buckets) append_archive(dir, bucket, evs);
    }
    // Drop the stale snapshot so the next read regenerates it (best-effort).
    std::error_code ec;
    fs::remove(paths::snapshot_path(dir), ec);

    CompactResult result{0, keep.size(), {}};
    for (const auto& [bucket, evs] : buckets) {
        result.archived += evs.size();
        result.archive_files.push_back(bucket);
    }
    report(result, options.json_output);
    return 0;
}

}  // namespace rp::compact

File: src/compact.rs
//! `rp compact` — archive terminal tickets' events and truncate the live log.
//!
//! Spec §6.1. Compaction is the one operation that rewrites the union-merged
//! `log.jsonl` instead of appending, so it is the one that can lose data if run
//! carelessly. It refuses unless the working tree is clean and `HEAD` is the
//! configured default branch (`--force` overrides; outside git there is nothing
//! to protect). It runs under the store lock so no appender races the rewrite,
//! and it appends to the archive *before* rewriting the log: an interruption
//! between the two steps leaves duplicated events (removed on read) rather
//! than lost ones.

use std::collections::BTreeMap;
use std::path::Path;

use crate::config::load_config;
use crate::error::{Error, Result, io_error};
use crate::events::Event;
use crate::fold::{TERMINAL, bare_id, fold};
use crate::json::{self, Json};
use crate::paths;
use crate::store;
use crate::time::{now_ms, parse_ts};
use crate::util::git_output;

/// Default retention before a terminal ticket's events are archived.
pub const DEFAULT_ARCHIVE_AFTER_DAYS: i64 = 90;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CompactResult {
    pub archived: usize,
    pub remaining: usize,
    pub archive_files: Vec<String>,
}

impl CompactResult {
    pub fn to_json(&self) -> Json {
        Json::Obj(vec![
            ("archived".into(), Json::Int(self.archived as i64)),
            ("remaining".into(), Json::Int(self.remaining as i64)),
            (
                "archive_files".into(),
                json::str_list(self.archive_files.iter().cloned()),
            ),
        ])
    }
}

/// `log-<YYYY>-Q<N>.jsonl` for an event timestamp (unparseable → `log-unknown.jsonl`).
fn quarter_bucket(ts: &str) -> String {
    match parse_ts(ts) {
        Some(ms) => {
            let stamp = crate::time::format_ts(ms.max(0) as u64);
            let year = &stamp[0..4];
            let month: u32 = stamp[5..7].parse().unwrap_or(1);
            format!("log-{year}-Q{}.jsonl", (month - 1) / 3 + 1)
        }
        None => "log-unknown.jsonl".to_string(),
    }
}

/// A refusal reason if compaction must not proceed, else `None`.
fn guard(repo_root: &Path, force: bool, default_branch: Option<&str>) -> Option<String> {
    if force {
        return None;
    }
    let inside_git = git_output(Some(repo_root), &["rev-parse", "--is-inside-work-tree"])
        .as_deref()
        == Some("true");
    if !inside_git {
        return None; // outside git: nothing to protect
    }
    let dirty =
        git_output(Some(repo_root), &["status", "--porcelain"]).is_some_and(|s| !s.is_empty());
    if dirty {
        return Some("refusing to compact: working tree is dirty (use --force to override)".into());
    }
    // An unborn branch (no commits yet) has no name; git reports HEAD.
    let branch = git_output(Some(repo_root), &["rev-parse", "--abbrev-ref", "HEAD"])
        .unwrap_or_else(|| "HEAD".into());
    let expected = default_branch.unwrap_or("main");
    if branch != expected {
        return Some(format!(
            "refusing to compact: HEAD is on '{branch}', not '{expected}' (use --force to override)"
        ));
    }
    None
}

fn is_terminal_set(ev: &Event) -> bool {
    ev.op == "set"
        && ev
            .set
            .as_ref()
            .and_then(|s| s.iter().find(|(k, _)| k.as_ref() == "status"))
            .and_then(|(_, v)| v.as_str())
            .is_some_and(|s| TERMINAL.contains(&s))
}

/// Run compaction. `Err(Error::Ticket)` carries a refusal; other errors are I/O.
pub fn run(rohrpost_dir: &Path, archive_after_days: i64, force: bool) -> Result<CompactResult> {
    let config = load_config(rohrpost_dir)?;
    if let Some(reason) = guard(
        paths::repo_root(rohrpost_dir),
        force,
        config.default_branch.as_deref(),
    ) {
        return Err(Error::Ticket(reason));
    }
    let cutoff_ms = now_ms() as i64 - archive_after_days * 86_400_000;

    // Read and partition while holding the same lock as the rewrite, so no
    // append between the read and the rewrite can be lost. Terminal-ness is
    // judged over everything (archive included); only live events move.
    let _guard = store::file_lock(rohrpost_dir)?;
    let events = store::read_events(rohrpost_dir)?;
    let live = store::read_live_events(rohrpost_dir)?;
    let by_id = fold(&events);

    // Tickets terminal since before the cutoff (judged by their last terminal event).
    let mut last_terminal: BTreeMap<String, i64> = BTreeMap::new();
    for ev in events.iter().filter(|e| is_terminal_set(e)) {
        if let Some(ms) = parse_ts(&ev.ts) {
            let slot = last_terminal
                .entry(bare_id(&ev.ticket).to_string())
                .or_insert(i64::MIN);
            *slot = (*slot).max(ms);
        }
    }
    let archivable: std::collections::HashSet<&str> = by_id
        .values()
        .filter(|t| TERMINAL.contains(&t.status.as_str()))
        .filter(|t| last_terminal.get(&t.id).is_some_and(|ms| *ms < cutoff_ms))
        .map(|t| t.id.as_str())
        .collect();

    let mut keep: Vec<&Event> = Vec::new();
    let mut buckets: BTreeMap<String, Vec<&Event>> = BTreeMap::new();
    for ev in &live {
        if archivable.contains(bare_id(&ev.ticket)) {
            buckets.entry(quarter_bucket(&ev.ts)).or_default().push(ev);
        } else {
            keep.push(ev);
        }
    }
    let by_ts_id = |a: &&Event, b: &&Event| {
        (a.ts.as_str(), a.id.as_str()).cmp(&(b.ts.as_str(), b.id.as_str()))
    };

    // Archive first (append-only, dedup-safe on read), then rewrite the log.
    let adir = paths::archive_dir(rohrpost_dir);
    std::fs::create_dir_all(&adir).map_err(|e| io_error("cannot create", &adir, &e))?;
    for (bucket, evs) in buckets.iter_mut() {
        evs.sort_by(by_ts_id);
        let target = adir.join(bucket);
        let payload: String = evs.iter().map(|e| e.encode() + "\n").collect();
        use std::io::Write as _;
        let mut file = std::fs::OpenOptions::new()
            .append(true)
            .create(true)
            .open(&target)
            .map_err(|e| io_error("cannot open", &target, &e))?;
        file.write_all(payload.as_bytes())
            .map_err(|e| io_error("cannot write", &target, &e))?;
    }
    keep.sort_by(by_ts_id);
    let payload: String = keep.iter().map(|e| e.encode() + "\n").collect();
    store::write_atomic(&paths::log_path(rohrpost_dir), payload.as_bytes())?;

    Ok(CompactResult {
        archived: buckets.values().map(Vec::len).sum(),
        remaining: keep.len(),
        archive_files: buckets.keys().cloned().collect(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn quarter_buckets_follow_the_event_timestamp() {
        assert_eq!(
            quarter_bucket("2026-02-01T00:00:00.000Z"),
            "log-2026-Q1.jsonl"
        );
        assert_eq!(
            quarter_bucket("2026-12-31T23:59:59.999Z"),
            "log-2026-Q4.jsonl"
        );
        assert_eq!(quarter_bucket("nonsense"), "log-unknown.jsonl");
    }
}
